src/renderers/shared/fiber/__tests__/ReactTopLevelFragment-test.js

```diff
@@ -94,4 +94,73 @@ describe('ReactTopLevelFragment', function() {
 
   });
 
+  it('preserves state if an implicit key slot switches from/to null', function() {
+
+    var instance = null;
+
+    class Stateful extends React.Component {
+      render() {
+        instance = this;
+        return <div>World</div>;
+      }
+    }
+
+    function Fragment({ condition }) {
+      return condition ? [null, <Stateful />] :
+        [<div>Hello</div>, <Stateful />];
+    }
+    ReactNoop.render(<Fragment />);
+    ReactNoop.flush();
+
+    var instanceA = instance;
+
+    expect(instanceA).not.toBe(null);
+
+    ReactNoop.render(<Fragment condition={true} />);
+    ReactNoop.flush();
+
+    var instanceB = instance;
+
+    expect(instanceB).toBe(instanceA);
+
+    ReactNoop.render(<Fragment condition={false} />);
+    ReactNoop.flush();
+
+    var instanceC = instance;
+
+    expect(instanceC === instanceA).toBe(true);
```

Comment:
Actually I can't be cause when it errors that stalls `jest`. I spent some time trying to figure out why but the debugger doesn't work on it. My best bet is that a message simply gets dropped which stalls the process waiting for a response that never arrives. cc @cpojer